packages/database/src/cachers/EntityHierarchyCacher.js

```diff
@@ -28,64 +26,77 @@ export class EntityHierarchyCacher {
     // makes it possible)
-    const alreadyCached = !!(await this.models.ancestorDescendantRelation.findOne({
+    const numberAlreadyCached = await this.models.ancestorDescendantRelation.count({
       entity_hierarchy_id: hierarchyId,
-      ancestor_id: parentId,
-    }));
-    if (alreadyCached) {
+      ancestor_id: parentIds,
+      generational_distance: 1,
+    });
+    if (numberAlreadyCached === parentIds.length) {
       return;
     }
 
-    // get the next generation of entities
-    const childIds = await this.getNextGeneration(hierarchyId, parentId);
+    // check whether next generation uses entity relation links, or should fall back to parent_id
+    const nextGenerationIsCanonical = await this.checkIfNextGenerationIsCanonical(
+      hierarchyId,
+      parentIds,
+    );
 
-    // if childIds is empty, we've made it to the leaf nodes
-    if (childIds.length === 0) {
-      return;
-    }
+    const childIdsToAncestorIds = {};
+    for (const parentId of parentIds) {
+      const ancestorIds = parentIdsToAncestorIds[parentId];
+      const childIds = nextGenerationIsCanonical
+        ? await this.getNextGenerationCanonically(parentId)
+        : await this.getNextGenerationViaEntityRelation(hierarchyId, parentId);
 
-    // cache this generation
-    const parentAndAncestorIds = [parentId, ...ancestorIds];
-    await this.cacheGeneration(hierarchyId, childIds, parentAndAncestorIds);
+      // if childIds is empty, we've made it to the leaf nodes
+      if (childIds.length === 0) {
+        continue;
+      }
+
+      // cache this generation
+      const parentAndAncestorIds = [parentId, ...ancestorIds];
+      await this.cacheGeneration(hierarchyId, childIds, parentAndAncestorIds);
+
+      // add this generation to the object for caching in the next recursion
+      childIds.forEach(childId => {
+        childIdsToAncestorIds[childId] = parentAndAncestorIds;
+      });
+    }
 
     // keep recursing through the hierarchy
-    const taskQueue = new AsyncTaskQueue(BATCH_SIZE);
-    const childTasks = childIds.map(childId =>
-      taskQueue.add(() =>
-        this.recursivelyFetchAndCacheDescendants(hierarchyId, childId, parentAndAncestorIds),
-      ),
-    );
-    await Promise.all(childTasks);
+    await this.fetchAndCacheDescendants(hierarchyId, childIdsToAncestorIds);
   }
 
-  async getNextGeneration(hierarchyId, entityId) {
-    // get any matching alternative hierarchy relationships leading out of these parents
+  async checkIfNextGenerationIsCanonical(hierarchyId, entityIds) {
     const hierarchyLinks = await this.models.entityRelation.find({
-      parent_id: entityId,
+      parent_id: entityIds,
       entity_hierarchy_id: hierarchyId,
     });
-    const childIds = hierarchyLinks.map(l => l.child_id);
-
-    if (childIds.length > 0) {
-      return childIds;
-    }
+    // if no entity relation links, the next gen must use the canonical "parent_id" links
+    return hierarchyLinks.length === 0;
```

Comment:
Using `.count` here could probably make this request faster

Reply:
True!